### ISSUE TITLE
Unable to install Mac OSX 10.10.2 - gyp: Call to './util/has_lib.sh pangocairo' returned exit status 0

### ISSUE THREAD

**Issue by thealscott** [Automattic/node-canvas#630]
I have installed pkg-config, pixman and cairo using Homebrew, all latest versions.

When I run `npm install canvas`, i get the following:

```
npm install canvas

> canvas@1.2.9 install /Library/WebServer/Documents/frontend-boilerplate/gulp-test/node_modules/canvas
> node-gyp rebuild

not found: ldconfig
gyp: Call to './util/has_lib.sh pangocairo' returned exit status 0. while trying to load binding.gyp
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:355:16)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Darwin 14.1.0
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Library/WebServer/Documents/frontend-boilerplate/gulp-test/node_modules/canvas
gyp ERR! node -v v4.1.0
gyp ERR! node-gyp -v v3.0.1
gyp ERR! not ok 
npm WARN EPEERINVALID bl@0.7.0 requires a peer of stream-browserify@* but none was installed.
npm WARN EPACKAGEJSON oak-tree@1.0.0 No repository field.
npm ERR! Darwin 14.1.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "canvas"
npm ERR! node v4.1.0
npm ERR! npm  v3.3.3
npm ERR! code ELIFECYCLE

npm ERR! canvas@1.2.9 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the canvas@1.2.9 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the canvas package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls canvas
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Library/WebServer/Documents/frontend-boilerplate/gulp-test/npm-debug.log
```

I checked out the repo and ran `npm install`, got the following error output:

```
npm install

> canvas@1.2.9 install /Library/WebServer/Documents/node-canvas
> node-gyp rebuild

  SOLINK_MODULE(target) Release/canvas-postbuild.node
ld: library not found for -lgcc_s.10.5
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Release/canvas-postbuild.node] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:270:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Darwin 14.1.0
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Library/WebServer/Documents/node-canvas
gyp ERR! node -v v4.1.0
gyp ERR! node-gyp -v v3.0.1
gyp ERR! not ok 

npm ERR! Darwin 14.1.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v4.1.0
npm ERR! npm  v3.3.3
npm ERR! code ELIFECYCLE
npm ERR! canvas@1.2.9 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the canvas@1.2.9 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the canvas package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls canvas
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Library/WebServer/Documents/node-canvas/npm-debug.log
```

I've dug around other open tickets and found errors that seem similar, but nothing quite matching this. Am out of ideas! Any help would be much appreciated.

Thanks,
Al

**Comment by LinusU** [comment]
Do you have the latest xcode installed?

**Comment by thealscott** [comment]
Hmm, I am on 6.4. My colleague is also using xcode 6.4 and he hasn't had any problems, so I didn't think it would be an issue. 

It seems I need to update to 10.10.4+ to get xcode 7, so I will try that.

**Comment by thealscott** [comment]
I have updated to 10.10.5 and xcode 7, but still getting the same errors.

**Comment by roncli** [comment]
Same issue on 10.10.5 and XCode 7.  I also have command line tools 7.0 (7A220).

Note that this doesn't start happening to me until `which` is installed.  `which@1.0.9` is installed with grunt, which is in my tool chain.

`npm install canvas` is successful.

`npm install which@1.0.9 canvas` fails.  Here is the relevant part of the log using -verbose:

```
gyp verb gyp gyp format was not specified; forcing "make"
gyp info spawn python
gyp info spawn args [ '/usr/local/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/roncli/dev/git/github/test/node_modules/canvas/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/local/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/roncli/.node-gyp/4.1.1/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/Users/roncli/.node-gyp/4.1.1',
gyp info spawn args   '-Dnode_gyp_dir=/usr/local/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=node.lib',
gyp info spawn args   '-Dmodule_root_dir=/Users/roncli/dev/git/github/test/node_modules/canvas',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.' ]
not found: ldconfig
gyp: Call to './util/has_lib.sh pangocairo' returned exit status 0. while trying to load binding.gyp
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:355:16)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Darwin 14.5.0
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/roncli/dev/git/github/test/node_modules/canvas
gyp ERR! node -v v4.1.1
gyp ERR! node-gyp -v v3.0.3
gyp ERR! not ok 
```

**Comment by LinusU** [comment]
@roncli Would probably be fixed by #628

**Comment by roncli** [comment]
I got to play with this more after upgrading to 10.11.  The issue still persists on OSX 10.11.1, XCode 7.1.1, command line tools 7.1.1 (7B1005), and canvas@1.3.1.  However, I can successfully workaround this by uninstalling grunt, then installing canvas, then reinstalling grunt.

**Comment by LinusU** [comment]
@roncli Thats very strange, have you also tried running `xcode-select --install`? I've had that fix the issue even though I already had xcode installed.

**Comment by roncli** [comment]
`xcode-select: error: command line tools are already installed, use "Software Update" to install updates`

**Comment by LinusU** [comment]
Hmm, the output doesn't make any sense at all.

> Call to './util/has_lib.sh pangocairo' returned exit status 0

This seems like a good thing...

Could you post your `npm-debug.log` to a gist and link here?

**Comment by joeyfromspace** [comment]
I am having this issue too with the same error message. However, I do not have grunt installed. The workaround that works for me is to install canvas by itself first (just npm install canvas) then follow it with an npm install to install the rest of the packages.

**Comment by roncli** [comment]
For me it wasn't `grunt` directly, it was one of its dependencies, `which`.

**Comment by ralphholzmann** [comment]
Just chiming in that I'm having the exact same error. I'm on OS X 10.11.1.

```bash
$ node --version
v4.2.1
$ npm --version
3.4.0
$ xcode-select --install
xcode-select: error: command line tools are already installed, use "Software Update" to install updates
$ npm install canvas

> canvas@1.3.4 install /Users/rholzmann/workspace/htt/www/node_modules/canvas
> node-gyp rebuild

not found: ldconfig
gyp: Call to './util/has_lib.sh pangocairo' returned exit status 0. while trying to load binding.gyp
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/Users/rholzmann/.nvm/versions/node/v4.2.1/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:355:16)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Darwin 15.0.0
gyp ERR! command "/Users/rholzmann/.nvm/versions/node/v4.2.1/bin/node" "/Users/rholzmann/.nvm/versions/node/v4.2.1/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/rholzmann/workspace/htt/www/node_modules/canvas
gyp ERR! node -v v4.2.1
gyp ERR! node-gyp -v v3.0.3
gyp ERR! not ok
npm ERR! Darwin 15.0.0
npm ERR! argv "/Users/rholzmann/.nvm/versions/node/v4.2.1/bin/node" "/Users/rholzmann/.nvm/versions/node/v4.2.1/bin/npm" "install" "canvas"
npm ERR! node v4.2.1
npm ERR! npm  v3.4.0
npm ERR! code ELIFECYCLE

npm ERR! canvas@1.3.4 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the canvas@1.3.4 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the canvas package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls canvas
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/rholzmann/workspace/htt/www/npm-debug.log
```

**Comment by ralphholzmann** [comment]
Update: if I clone the `./has_lib.sh` file and run `./has_lib pangocairo` from my terminal, it returns true. For some reason in the context of npm this is failing.

**Comment by petershaw** [comment]
When i install canvas first, and than do the rest of my nom install it works, but if i install my modules and than add canvas it fails with the above error code. 
That's strange.

**Comment by ralphholzmann** [comment]
One more update here: there's definitely something up with `./has_lib.sh` on OS X. I cloned the repo and hard coded `has_lib.sh` to always return true (`1`). Everything installed correctly and runs great.

**Comment by niftylettuce** [comment]
Can this get PR'd?

**Comment by calidion** [comment]
+1 for this problem.

**Comment by majamatic** [comment]
+1

**Comment by LinusU** [comment]
#701 should fix this. @niftylettuce, @calidion and @majamatic could you please test that?

**Comment by majamatic** [comment]
I am now getting this:

`> canvas@1.0.4 install /Users/maya/projects/node_modules/grunt-spritely/node_modules/canvas
> node-gyp rebuild

not found: ldconfig
gyp: Call to './util/has_lib.sh jpeg' returned exit status 0 while in binding.gyp. while trying to load binding.gyp
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:305:16)
gyp ERR! stack     at ChildProcess.emit (events.js:110:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1074:12)
gyp ERR! System Darwin 15.2.0
gyp ERR! command "node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/maya/projects/node_modules/grunt-spritely/node_modules/canvas
gyp ERR! node -v v0.12.7
gyp ERR! node-gyp -v v3.2.1
gyp ERR! not ok 
npm WARN install:canvas@1.0.4 canvas@1.0.4 install: `node-gyp rebuild`
npm WARN install:canvas@1.0.4 Exit status 1

> canvas@1.3.6 install /Users/maya/projects/node_modules/canvas`

**Comment by majamatic** [comment]
The issue was fixed for me when I uninstalled grunt and installed it again. Thank you for the help

**Comment by simonrobb** [comment]
Another data point - reinstalling grunt didn't help me, but clearing out my node_modules and installing canvas before the other modules did, which agrees with what @petershaw and @roncli have commented.

**Comment by ooskapenaar** [comment]
@LinusU : that commit fixes my problems.  Nice cleanup.

**Comment by LinusU** [comment]
Thank you, I think it breaks in some cases on OS X though so I need to switch out `which` for a replacement. On which OS are you?

**Comment by ooskapenaar** [comment]
@LinusU, sorry, yesterday it worked fine once, now it is failing. My solution must have been cached by npm or something.
I am on Mac OS X 10.11.2, node 5.4.1, npm 3.3.12. Python 2.7. Using MacPorts.
@ralphholzmann has got the cause of this thing - has_lib returns true from bash, but somehow the return in the function is grabbed and taken as an exit signal when running through node-gyp.  Anyone else heard of that?  It sounds like node-gyp may be interfering with the way shell commands are run on Mac.
`gyp: Call to './util/has_lib.sh pangocairo' returned exit status 0. while trying to load binding.gyp`

**Comment by LinusU** [comment]
I belive that the problem is that `node-gyp` considers any output on `stderr` to be an error. Since `if which x; then` outputs `cannot find x` to stderr it fails...

**Comment by targos** [comment]
If this is a problem with the `which` call, I suggest you check if you have an old version (< 1.2.0) of the [`which`](https://www.npmjs.com/package/which) module in your `node_modules` directory.
I had that problem a few months ago because of one of my dependencies. Upgrading `which` fixed it.

**Comment by LinusU** [comment]
I think that we are currently using the system provided `which`, there are much better alternatives though. 

http://stackoverflow.com/questions/592620/check-if-a-program-exists-from-a-bash-script

**Comment by targos** [comment]
Well the problem is that if the `which` module is installed, it takes precedence over the system one (because of npm)

**Comment by targos** [comment]
I agree that `which` is not the best option. `command -v` seem a good alternative. Or just try to run `ldconfig --version` ?

**Comment by ooskapenaar** [comment]
I think running `ldconfig` when it's not defined on mac will also produce the same result (output on stderr).

**Comment by LinusU** [comment]
I've fixed the script, could you please help me test it? :+1: #701

**Comment by ooskapenaar** [comment]
I just tested this, and it works fine:
```
--- a/util/has_lib.sh
+++ b/util/has_lib.sh
@@ -7,7 +7,7 @@ has_lib() {
   PATH="$PATH:/sbin"
   export PATH
   # Try using ldconfig on linux systems
-  for LINE in `which ldconfig > /dev/null && ldconfig -p 2>/dev/null | grep -E $regex`; do
+  for LINE in `which ldconfig 2>&1 > /dev/null && ldconfig -p 2>/dev/null | grep -E $regex`; do
     return 0
   done
```

**Comment by targos** [comment]
I can try on Fedora

**Comment by ooskapenaar** [comment]
@LinusU I just cloned your repo and I am not seeing any recent changes. What do you mean you have fixed the script?

**Comment by LinusU** [comment]
@ooskapenaar You can try running `npm install LinusU/node-canvas#build-cleanup` in any directory, than it will try to install that version into `node_modules/canvas`

**Comment by ooskapenaar** [comment]
@LinusU: ok, didn't know about that branch.  That is very easy to read and it works just great.

+1  for Mac OS X

**Comment by LinusU** [comment]
This should be fixed by #701, released as `1.3.8`
